docs/feature-gates.md

```diff
@@ -70,35 +71,54 @@ NetworkPolicy implementation for Pod-to-Service traffic.
+opposed to >= 4.4 without this feature).
+
 ### EndpointSlice
 
 `EndpointSlice` enables Service EndpointSlice support in AntreaProxy. The EndpointSlice API was introduced in Kubernetes
-1.16 (alpha) and it is enabled by default in Kubernetes 1.17 (beta). The EndpointSlice feature gate will take no effect
-if AntreaProxy is not enabled. The endpoint conditions of `Serving` and
-`Terminating` are not supported currently. ServiceTopology is not supported either. Refer to
-this [link](https://kubernetes.io/docs/tasks/administer-cluster/enabling-endpointslices/)
-for more information. The EndpointSlice API version that AntreaProxy supports is v1beta1 currently, and other
-EndpointSlice API versions are not supported. If EndpointSlice is enabled in AntreaProxy, but EndpointSlice API is
-disabled in Kubernetes or EndpointSlice API version v1beta1 is not supported in Kubernetes, Antrea Agent will log an
-error message and will not implement Cluster IP functionality as expected.
+1.16 (alpha) and it is enabled by default in Kubernetes 1.17 (beta), promoted to GA in Kubernetes 1.21. The EndpointSlice
+feature will take no effect if AntreaProxy is not enabled. Refer to this [link](https://kubernetes.io/docs/tasks/administer-cluster/enabling-endpointslices/)
+for more information about EndpointSlice. Don't enable this feature if Kubernetes version is lower than 1.21 or EndpointSlice
+API is disabled in Kubernetes, otherwise Antrea Agent will log an error message and will not implement Cluster IP functionality
+as expected.
 
 #### Requirements for this Feature
 
-When using the OVS built-in kernel module (which is the most common case), your kernel version must be >= 4.6 (as
-opposed to >= 4.4 without this feature).
+- Kubernetes version: >= 1.21.
+- EndpointSlice API is enabled in Kubernetes.
+- `AntreaProxy` is enabled.
 
 ### TopologyAwareHints
 
-`TopologyAwareHints` enables TopologyAwareHints support in AntreaProxy. The feature
-TopologyAwareHints is at beta stage in Kubernetes 1.23 (beta), and it is enabled by
-default in Kubernetes 1.24. For AntreaProxy, traffic can be routed to the Endpoint
-which is closer to its origin with this feature. Refer to this
-[link](https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/)
-for more information.
+`TopologyAwareHints` enables TopologyAwareHints support in AntreaProxy. For AntreaProxy, traffic can be routed to the
+Endpoint which is closer to where it originated when this feature is enabled. Refer to this [link](https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/)
+for more information about TopologyAwareHints.
+
+#### Requirements for this Feature
+
+- `AntreaProxy` is enabled.
+- `EndpointSlice` is enabled.
+
+### ProxyTerminatingEndpoints
+
+`ProxyTerminatingEndpoints` enables ProxyTerminatingEndpoints support in AntreaProxy. When ProxyTerminatingEndpoints is
```

Comment:
Question: if the support is controlled by feature gate and disabled by default, could Antrea pass latest K8s conformance test? I remember you mentioned latest conformance requires some EndpointSlice related feature to be supported.

Reply:
I'll verify that.

Reply:
cc @jianjuns for input

Reply:
The feature gate is inherited from K8s? But sure if we believe the risk is low we can remove it or enable by default.

Reply:
Yes, it's inherited from kube-proxy.

Reply:
> Then if we merge it as is, the test case won't pass unless the ProxyTerminatingEndpoints feature is enabled?

If we want to pass corresponding tests about ProxyTerminatingEndpoints, we need to enable both this feature and proxyAll.

I think we might need a feature gate since like TopologyAwareHints in AntreaProxy has a feature gateway, ProxyTerminatingEndpoints should be consistent with that.

Reply:
> If we want to pass corresponding sig-network tests about ProxyTerminatingEndpoints, we need to enable both this feature and proxyAll.

Why? If proxyAll is not running, kube-proxy must be running, right? How could kube-proxy fail the test?

> I think we might need a feature gate since like TopologyAwareHints in AntreaProxy has a feature gateway, ProxyTerminatingEndpoints should be consistent with that.

I don't feel such consistency is meaningful. We do have many small features that are added directly, like you did in https://github.com/antrea-io/antrea/pull/2792. The rule to add a feature gate should be related to the code/feature's risk and maturity, instead of whether there is one in K8s.

Reply:
> Why? If proxyAll is not running, kube-proxy must be running, right? How could kube-proxy fail the test?

Sorry, I missed something: as you said, the tests will always pass with kube-proxy. The code of TopologyAwareHints in AntreaProxy when proxyAll is enabled takes effect.

> I don't feel such consistency is meaningful. We do have many small features that are added directly, like you did in https://github.com/antrea-io/antrea/pull/2792. The rule to add a feature gate should be related to the code/feature's risk and maturity, instead of whether there is one in K8s.

If so, do you think we could also remove the feature gate TopologyAwareHints in Antrea?

Reply:
Then there should be no problem if we just remove ProxyTerminatingEndpoints featuregate regardless of the configuration?

I didn't evaluate the risk of TopologyAwareHints but at least it's worth to change what has been added, which would just cause more confusion. We should consider promoting it to Beta and GA instead according to it's maturity.

Reply:
I don't think we have anymore problem if we just remove ProxyTerminatingEndpoints. Then with proxyAll enabled and without kube-proxy, AntreaProxy can pass the follow sig-network test cases:

```
[It] [sig-network] Services should fallback to terminating endpoints when there are no ready endpoints with internalTrafficPolicy=Cluster [Feature:ProxyTerminatingEndpoints]
[It] [sig-network] Services should fallback to local terminating endpoints when there are no ready endpoints with internalTrafficPolicy=Local [Feature:ProxyTerminatingEndpoints]
[It] [sig-network] Services should fallback to terminating endpoints when there are no ready endpoints with externallTrafficPolicy=Cluster [Feature:ProxyTerminatingEndpoints]
[It] [sig-network] Services should fail health check node port if there are only terminating endpoints [Feature:ProxyTerminatingEndpoints]
```

The test that Antrea cannot pass is:
```
[It] [sig-network] Services should fallback to local terminating endpoints when there are no ready endpoints with externalTrafficPolicy=Local [Feature:ProxyTerminatingEndpoints]
```

The root cause is that AntreaProxy doesn't select any remote Endpoints as backend when externalTrafficPolicy is Local (in this test, no local Endpoint is available in a sub test), while kube-proxy could select remote Endpoints as backend when externalTrafficPolicy is Local and client is from in-cluster. I'll add another PR to support this later.